### PR TITLE
feat: Copy All Drafts **except Form** Into Draft Packages

### PR DESCRIPTION
**Pull request by mbylstra**
### List of things to do manually once Seb's branch has been merged:

- run the python script (sitting on my laptop at the moment) and push to CI - ✅ 
- where there are dependencies on other drafts, point at the old drafts rather than the new ones. This includes tests (this is why tests are failing). Look at failing build. ✅ 
- manually add 3rd party deps (look at the failing build) - ✅ 
- Lint MOSTLY DONE, just superficial stuff  ✅ 
- Lint fix the new package.json files ✅ 
- Moai ✅ 
- make sure the lerna build is working ? NOT YET

### Unknowns
- will we have two copies of stories - is this an issue? Do we need to remove the stories from the old drafts? Can do manually easy enough.

### Some decisions made

- Unlike Well, I kept the tests and stories inside the `KaizenDraft/ComponentName` dir rather than at the root of the package. This was easier to do, the tests still pass (they should!), and I think it's easier to find the tests when you are working on some source code. To keep things consistent, I think the Well tests/stories can be move later.

### Out of scope:
- removing the old draft packages inside the core component library package. We will do that once the other teams have run the codemods.
- there's an opportunity to remove no longer used deps from the component library package. This can be done in a separate PR considering time constraints.
- The package.json description will have names like "Vertical progress step" instead of "Vertical Progress Step". Can fix this up later if we want
- The .elm files are still pointing at the old draft scss files. Eg: `@kaizen/component-library/draft/Kaizen/Tooltip/Tooltip.scss`. This is dodgy, but we _can_ fix it later. I'd prefer to descope this.
- removing the duplicated stories from the old draft directory

### The python script (warning big)

```
import os
import shutil
from inflection import *


os.chdir("../kaizen-design-system")

old_draft_dir = "./packages/component-library/draft/Kaizen"


def get_new_source_dir(package_dir, dir_name):
    return "{}/KaizenDraft/{}".format(package_dir, dir_name)


def get_new_package_dir(dashed_name):
    return "./draft-packages/{}".format(dashed_name)


def get_old_source_dir(component_dir):
    return "{}/{}".format(old_draft_dir, component_dir)


def make_metadata(dir_name):
    dashed_name = dasherize(underscore(dir_name))
    human_name = humanize(underscore(dir_name))
    package_dir = get_new_package_dir(dashed_name)
    return {
        "package_dir_name": dashed_name,
        "package_dir": package_dir,
        "package_json_path": "{}/package.json".format(package_dir),
        "old_source_dir": get_old_source_dir(dir_name),
        "package_relative_source_dir": "./KaizenDraft/{}".format(dir_name),
        "new_source_dir": get_new_source_dir(package_dir, dir_name),
        "npm_package_name": "@kaizen/draft-{}".format(dashed_name),
        "npm_description": "The draft {} component".format(human_name),
        "tsconfig_template_path": "./draft-packages/well/tsconfig.dist.json",
    }


def get_package_json_source(metadata):
    return """
{
  "name": "%s",
  "version": "1.0.0",
  "description": "%s",
  "scripts": {
    "prepublish": "tsc --project tsconfig.dist.json",
    "build": "yarn clean && yarn prepublish",
    "build:watch": "yarn clean && yarn prepublish --watch",
    "clean": "rimraf '*.d.ts' '*.js' '*.map'"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/cultureamp/kaizen-design-system.git",
    "directory": "draft-packages/%s"
  },
  "bugs": {
    "url": "https://github.com/cultureamp/kaizen-design-system/issues"
  },
  "files": [
    "**/*",
    "!**/*.ts",
    "!**/*.tsx",
    "**/*.d.ts",
    "!**/*.spec.*",
    "!**/*.snap",
    "!draft-templates",
    "!stories",
    "!tsconfig.dist.json"
  ],
  "author": "",
  "private": false,
  "license": "MIT",
  "dependencies": {
    "@kaizen/component-library": "^7.6.1",
    "@types/classnames": "^2.2.10",
    "classnames": "^2.2.6"
  },
  "devDependencies": {
    "@cultureamp/elm-storybook": "cultureamp/elm-storybook#0.1.0",
    "rimraf": "^3.0.2"
  },
  "peerDependencies": {
    "@kaizen/design-tokens": "1.2.0",
    "react": "^16.9.0"
  }
}""" % (
        metadata["npm_package_name"],
        metadata["npm_description"],
        metadata["package_dir_name"],
    )


def make_new_package_dir(metadata):
    try:
        os.mkdir(metadata["package_dir"])
    except:
        pass


def make_package_json(metadata):
    source = get_package_json_source(metadata)
    with open(metadata["package_json_path"], "w") as file:
        file.write(source)


def copy_source_files(metadata):
    try:
        shutil.copytree(metadata["old_source_dir"], metadata["new_source_dir"])
    except:
        pass


def copy_tsconfig(metadata):
    try:
        shutil.copyfile(
            metadata["tsconfig_template_path"], metadata["package_dir"] + "/tsconfig.dist.json"
        )
    except:
        pass


def create_tsx_barrel_file(metadata):
    content = 'export * from "{}"'.format(metadata["package_relative_source_dir"])
    path = "{}/index.ts".format(metadata["package_dir"])
    with open(path, "w") as f:
        f.write(content)


old_draft_dirs = os.listdir(old_draft_dir)
old_draft_dirs.remove("Form")  # do everything except for Form
old_draft_dirs.remove("Well")  # well has already been done
for component_dir in old_draft_dirs:
    metadata = make_metadata(component_dir)
    make_new_package_dir(metadata)
    make_package_json(metadata)
    copy_source_files(metadata)
    copy_tsconfig(metadata)
    create_tsx_barrel_file(metadata)
```
So, this does everything except for:

    adding 3rd party deps to package.json files
    updating references to other draft packages (in source and and in tests)

Needs `pip install inflection` for the dash-casing stuff.

Also there is a little bash script for updating the Elm module names:

```fastmod "module (Kaizen)\." "module KaizenDraft.${1}" ../kaizen-design-system/draft-packages/```

(`brew install fastmod` first)

### CODE REVIEW

**Comment by mbylstra** [comment]
Still to do:

- move test files to the root of the package ?
- create barrel files for tsx
- Update all the Elm module names inside source code so that they are `KaizenDraft` instead of `Draft`

**Comment by mbylstra** [comment]
[Copy All Drafts **except Form** Into Draft Packages](https://trello.com/c/S8f0qlYe/904-copy-all-drafts-except-form-into-draft-packages)

**Comment by mbylstra** [comment]
There are some more things that will need to be done that I don't think will be easy to automate. Eg: some of the packages have 3rd party deps that need to be added to package.json. My plan is to do this stuff manually, after Seb's branch has been merged.

**Comment by mbylstra** [comment]
The python script. The reason it's not in the PR is that I want to be able to do `git clean -f` to get rid of changes and not accidentally delete my script!

```
import os
import shutil
from inflection import *


os.chdir("../kaizen-design-system")

old_draft_dir = "./packages/component-library/draft/Kaizen"


def get_new_source_dir(package_dir, dir_name):
    return "{}/KaizenDraft/{}".format(package_dir, dir_name)


def get_new_package_dir(dashed_name):
    return "./draft-packages/{}".format(dashed_name)


def get_old_source_dir(component_dir):
    return "{}/{}".format(old_draft_dir, component_dir)


def make_metadata(dir_name):
    dashed_name = dasherize(underscore(dir_name))
    human_name = humanize(underscore(dir_name))
    package_dir = get_new_package_dir(dashed_name)
    return {
        "package_dir_name": dashed_name,
        "package_dir": package_dir,
        "package_json_path": "{}/package.json".format(package_dir),
        "old_source_dir": get_old_source_dir(dir_name),
        "package_relative_source_dir": "./KaizenDraft/{}".format(dir_name),
        "new_source_dir": get_new_source_dir(package_dir, dir_name),
        "npm_package_name": "@kaizen/draft-{}".format(dashed_name),
        "npm_description": "The draft {} component".format(human_name),
        "tsconfig_template_path": "./draft-packages/well/tsconfig.dist.json",
    }


def get_package_json_source(metadata):
    return """
{
  "name": "%s",
  "version": "1.0.0",
  "description": "%s",
  "scripts": {
    "prepublish": "tsc --project tsconfig.dist.json",
    "build": "yarn clean && yarn prepublish",
    "build:watch": "yarn clean && yarn prepublish --watch",
    "clean": "rimraf '*.d.ts' '*.js' '*.map'"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/cultureamp/kaizen-design-system.git",
    "directory": "draft-packages/well"
  },
  "bugs": {
    "url": "https://github.com/cultureamp/kaizen-design-system/issues"
  },
  "files": [
    "**/*",
    "!**/*.ts",
    "!**/*.tsx",
    "**/*.d.ts",
    "!**/*.spec.*",
    "!**/*.snap",
    "!draft-templates",
    "!stories",
    "!tsconfig.dist.json"
  ],
  "author": "",
  "private": false,
  "license": "MIT",
  "dependencies": {
    "@kaizen/component-library": "^7.6.1",
    "@types/classnames": "^2.2.10",
    "classnames": "^2.2.6"
  },
  "devDependencies": {
    "@cultureamp/elm-storybook": "cultureamp/elm-storybook#0.1.0",
    "rimraf": "^3.0.2"
  },
  "peerDependencies": {
    "@kaizen/design-tokens": "1.2.0",
    "react": "^16.9.0"
  }
}""" % (
        metadata["npm_package_name"],
        metadata["npm_description"],
    )


def make_new_package_dir(metadata):
    try:
        os.mkdir(metadata["package_dir"])
    except:
        pass


def make_package_json(metadata):
    source = get_package_json_source(metadata)
    with open(metadata["package_json_path"], "w") as file:
        file.write(source)


def copy_source_files(metadata):
    try:
        shutil.copytree(metadata["old_source_dir"], metadata["new_source_dir"])
    except:
        pass


def copy_tsconfig(metadata):
    try:
        shutil.copyfile(
            metadata["tsconfig_template_path"], metadata["package_dir"] + "/tsconfig.dist.json"
        )
    except:
        pass


def create_tsx_barrel_file(metadata):
    content = 'export * from "{}"'.format(metadata["package_relative_source_dir"])
    path = "{}/index.ts".format(metadata["package_dir"])
    with open(path, "w") as f:
        f.write(content)


old_draft_dirs = os.listdir(old_draft_dir)
old_draft_dirs.remove("Form")  # do everything except for Form
old_draft_dirs.remove("Well")  # well has already been done
for component_dir in old_draft_dirs:
    metadata = make_metadata(component_dir)
    make_new_package_dir(metadata)
    make_package_json(metadata)
    copy_source_files(metadata)
    copy_tsconfig(metadata)
    create_tsx_barrel_file(metadata)
```
So, this does everything except for:
- adding 3rd party deps to package.json files
- updating references to other draft packages (in source and and in tests)

**Comment by mbylstra** [comment]
Found a bug: the repository/directory in the package.json is still `well`. Will fix.

**Comment by mbylstra** [comment]
This has now been merged with Seb's PR. 

I realised I made a mistake in the barrel file generation... looking into it.

**Comment by ActuallyACat** [comment]
> will we have two copies of stories - is this an issue? Do we need to remove the stories from the old drafts? Can do manually easy enough.

This is important, but not important enough to do for launch. Let's get this out the door and I can clean up the stories after we've rolled this out.

> Unlike Well, I kept the tests and stories inside the KaizenDraft/ComponentName dir rather than at the root of the package. This was easier to do, the tests still pass (they should!), and I think it's easier to find the tests when you are working on some source code. To keep things consistent, I think the Well tests/stories can be move later.

😬 Erm so. I went in another direction. I have kept `KaizenDraft` dir but only for elm, everything else is flat (like the well package). I can change mine since I have less files to cover.

> removing the old draft packages inside the core component library package. We will do that once the other teams have run the codemods.
✅

> there's an opportunity to remove no longer used deps from the component library package. This can be done in a separate PR considering time constraints.
✅✅✅

> The package.json description will have names like "Vertical progress step" instead of "Vertical Progress Step". Can fix this up later if we want
✅

> The .elm files are still pointing at the old draft scss files. Eg: @kaizen/component-library/draft/Kaizen/Tooltip/Tooltip.scss. This is dodgy, but we can fix it later. I'd prefer to descope this.
✅ Agree on this. We can't add dependencies to files that haven't been released yet.

**Comment by mbylstra** [comment]
> 😬 Erm so. I went in another direction. I have kept KaizenDraft dir but only for elm, everything else is flat (like the well package). I can change mine since I have less files to cover.
Thanks, sorry about that! I didn't necessarily disagree with the folder structure, it was just harder to automate than just moving the existing folder structure with all source in the same folder.

**Comment by mbylstra** [comment]
Is `ci` the right tag for this PR?

**Comment by sebpearce** [comment]
I noticed that things like `draft-packages/title-block/KaizenDraft/TitleBlock/NavigationButtons.js.map` are part of the diff. Is this intentional i.e. we want to copy across the results of `yarn build`?

**Comment by mbylstra** [comment]
> I noticed that things like draft-packages/title-block/KaizenDraft/TitleBlock/NavigationButtons.js.map are part of the diff. Is this intentional i.e. we want to copy across the results of yarn build?

That's a mistake. Good pick up! I will remove. I think it was because i did a `yarn build` for testing.

**Comment by mbylstra** [comment]
I removed the `*.js.map files`

File: draft-packages/collapsible/package.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "@kaizen/draft-collapsible",
+  "version": "1.0.0",
+  "description": "The draft Collapsible component",
+  "scripts": {
+    "prepublish": "tsc --project tsconfig.dist.json",
+    "build": "yarn clean && yarn prepublish",
+    "build:watch": "yarn clean && yarn prepublish --watch",
+    "clean": "rimraf '*.d.ts' '*.js' '*.map'"
```
Comment (ActuallyACat) [inline draft-packages/collapsible/package.json:9]: The only thing to be careful of is this won't clean those `KaizenDraft/thing.js` files... Did you manually remove them when you made your last commit?

Comment (ActuallyACat) [reply draft-packages/collapsible/package.json:9]: Changing all of these to 
```
"clean": "rimraf '*.d.ts' '*.js' '*.map' 'KaizenDraft/**/*.d.ts' 'KaizenDraft/**/*.js' 'KaizenDraft/**/*.map'"
```
should do the trick

**Comment by sebpearce** [comment]
Same here. Ready 2 rumble

**Comment by mbylstra** [comment]
> Just clean up the clean script and this looks good to me. Nice work!
Done in https://github.com/cultureamp/kaizen-design-system/pull/431/commits/6f40a6f1432b0080f572e6df24ba20063fbe8772

**Pull request merged**